# MIPROv2

Companion notebook for Chapter 6, "MIPROv2" (pages 143–144).

Jointly propose instructions and demonstrations, then use validation feedback to search their combinations.

**Use it when:** Both prompt wording and examples may matter and you can spend more compile calls on a joint search.

**What compilation changes:** Instruction plus up to two bootstrapped and two labeled demonstrations.

Important configuration in this benchmark:

- `auto="light"` search budget
- up to two bootstrapped and two labeled demonstrations
- GPT-5.6-sol proposes instructions; GPT-5.6-luna runs the task

Every notebook in this folder loads the chapter's 300-passage benchmark
and its frozen, pair-grouped split: 160 training, 60 validation, and 80
locked-test examples. Both members of a human/AI pair stay in the same
partition. Optimizer choices use validation only; the locked test is
scored once, after the program is frozen. These scores compare optimizer
behavior; they are not a general AI-detector leaderboard.

**Keys and cost:** The default **Run All** is free: it needs no API key, makes no model call, and shows the saved result of the recorded run. A live run (`CHAPTER06_RUN_LIVE=1`) makes paid model calls and needs `OPENAI_API_KEY`. The recorded run cost $0.8548 to optimize, which took 285.2 seconds (4.8 minutes), and $0.2138 to evaluate. A live run is a new experiment, so its score, cost, and time will differ.

In [1]:
import os
import sys
from pathlib import Path

import dspy
from dotenv import load_dotenv

cwd = Path.cwd().resolve()
REPO_ROOT = cwd if (cwd / "chapter06").is_dir() else cwd.parent
if not (REPO_ROOT / "chapter06" / "results" / "expanded_notebooks" / "comparison.json").exists():
    raise RuntimeError("Run this notebook from the repository root or the chapter06 directory.")
if str(REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(REPO_ROOT))
load_dotenv(REPO_ROOT / ".env")

from chapter06.notebook_support import artifact_paths, learned_program_preview, verify_prompt_artifact
from chapter06.optimizer_runtime import format_result, published_result, score_summary

OPTIMIZER = 'miprov2'
RUN_LIVE = os.getenv("CHAPTER06_RUN_LIVE", "0") == "1"
if RUN_LIVE and not os.getenv("OPENAI_API_KEY"):
    raise RuntimeError(
        "CHAPTER06_RUN_LIVE=1 needs OPENAI_API_KEY. Add it to the .env file "
        "in the repository root and restart the kernel."
    )
print(f"optimizer={OPTIMIZER!r}; live={RUN_LIVE}")

optimizer='miprov2'; live=False


## Load the frozen dataset

The next cell is the setup listing from Chapter 6, "Setting Up the AI Detection
Module" (page 135). Every notebook in this folder starts with it.

Cells that call a model API hold the chapter listing unchanged, indented under
`if RUN_LIVE:`. By default they are skipped, so **Run All** is free and works
offline, and the saved result of the recorded run is displayed instead. To run
the listings yourself, set `CHAPTER06_RUN_LIVE=1` before starting Jupyter.
Every live run needs `OPENAI_API_KEY` in the repository `.env` file; if the
key is missing, the setup cell above stops with a message.

In [2]:
# Load the frozen Chapter 6 dataset and pair-grouped split
from chapter06.optimizer_runtime import (
    AIDetector,
    exact_match,
    hashed_ngram_embeddings,
    load_frozen_examples,
    split_summary,
)

NUM_THREADS = 1
splits = load_frozen_examples()
trainset = splits["train"]
valset = splits["validation"]
testset = splits["test"]

print(split_summary(splits))

train=160 (human=80, AI=80); validation=60 (human=30, AI=30); test=80 (human=40, AI=40)


> **Book vs. notebook:** The chapter creates `task_lm` in its baseline listing (page 135) and uses `teacher_lm` without printing its definition. The cell below defines both so the listing that follows runs as printed: GPT-5.6-luna runs the task and GPT-5.6-sol is the teacher (page 157). Creating the model objects makes no API call.

In [3]:
task_lm = dspy.LM("openai/gpt-5.6-luna", cache=False)
teacher_lm = dspy.LM("openai/gpt-5.6-sol", cache=False)
dspy.configure(lm=task_lm)
print(f"task model: {task_lm.model}")
print(f"teacher model: {teacher_lm.model}")

task model: openai/gpt-5.6-luna
teacher model: openai/gpt-5.6-sol


## Compile with MIPROv2

Page 143. The light preset still runs a full instruction and demonstration search against the validation split.

In [4]:
if RUN_LIVE:
    optimizer = dspy.MIPROv2(
            metric=exact_match,
            prompt_model=teacher_lm,
            task_model=task_lm,
            auto="light",  # Use light preset for faster benchmarking
            num_threads=NUM_THREADS,
            max_bootstrapped_demos=2,
            max_labeled_demos=2,
            track_stats=True,
        )
    optimized_detector = optimizer.compile(
        AIDetector(),
        trainset=trainset,
        valset=valset,
    )
else:
    print(
        "Skipped: this listing calls the model API. Set "
        "CHAPTER06_RUN_LIVE=1 before starting Jupyter to run it; the saved "
        "result is shown below."
    )

Skipped: this listing calls the model API. Set CHAPTER06_RUN_LIVE=1 before starting Jupyter to run it; the saved result is shown below.


## Score the compiled program

In live mode the next cell scores `optimized_detector`, the program the listing above produced, on the validation split and then once on the locked test. The cell first prints the instruction your live run selected and the number of demonstrations it attached. Every prediction is a fresh model call: the 60 validation and 80 locked-test predictions run one at a time, which takes five to ten minutes, depending on the instruction. The summary lines are printed when both passes finish.

A call that fails, or an answer that cannot be read as true or false, counts as a wrong prediction; the summary then ends with a `Failed Calls` line. If the first three calls fail outright, or no validation prediction can be scored, the cell stops and shows the first error.

In [5]:
if RUN_LIVE:
    predictor = optimized_detector.detect.predict
    print(f"Instruction: {predictor.signature.instructions}")
    print(f"Demonstrations: {len(predictor.demos)}\n")
    print(score_summary(optimized_detector, valset, testset))
else:
    print(
        "Skipped: scoring calls the model. Set CHAPTER06_RUN_LIVE=1 before "
        "starting Jupyter to run it; the saved result is shown below."
    )

Skipped: scoring calls the model. Set CHAPTER06_RUN_LIVE=1 before starting Jupyter to run it; the saved result is shown below.


## Saved result

The block below is the recorded run for this notebook, in the layout of the result blocks printed in the chapter. Every notebook uses the same layout, so a block can show details the chapter prints for some optimizers only: the count behind the validation percentage, token totals, the combined cost, the time in minutes, and percentages to two decimal places. It is read from `results/expanded_notebooks/comparison.json`, so it is the same whether or not you run live. The recorded run was produced by `run_optimizer()` in `chapter06/optimizer_runtime.py`.

In [6]:
print(format_result(published_result(OPTIMIZER)))
print()
print(artifact_paths(OPTIMIZER))

OPTIMIZER: MIPROv2
Validation Accuracy:   80.00% (48/60)
Locked Test Accuracy:  66.25% (53/80)
Same-Model Baseline:   53.75% (43/80)
Accuracy Uplift:       +12.50 percentage points
------------------------------------------------------------
Total Tokens:          565,949
  - Prompt:            471,085
  - Completion:        94,864
Optimization Cost:     $0.8548
Evaluation Cost:       $0.2138
Total Recorded Cost:   $1.0685
Optimization Time:     285.2s (4.8 minutes)
Mean / p95 Latency:    2.098s / 2.870s

Saved artifacts:
- program snapshot: chapter06/results/expanded_notebooks/miprov2/full/optimized_program.json
- prompt snapshot: chapter06/results/expanded_notebooks/miprov2/full/learned_prompt.json
- chapter comparison: chapter06/CHAPTER_RESULTS.md


**Output printed in the book (page 144):**

```text
============================================================
OPTIMIZER: MIPROv2
============================================================
Validation Accuracy: 80.0% (48/60)
Locked Test Accuracy: 66.3% (53/80)
Same-Model Baseline:  53.8% (43/80)
Accuracy Uplift:      +12.5 percentage points
------------------------------------------------------------
Total Tokens:         565,949
  - Prompt:           471,085
  - Completion:       94,864
Optimization Cost:    $0.8548
Evaluation Cost:      $0.2138
Total Recorded Cost:  $1.0685
Optimization Time:    285.2s (4.8 minutes)
Mean / p95 Latency:   2.098s / 2.870s
============================================================
```

These are the numbers of the saved result above, which shows the percentages to two decimal places; a live run is a new experiment, so its scores, cost, and timings will differ.

## Read the result

MIPROv2 searches instructions and demonstration sets together, so read the two as one result. The saved program pairs a rewritten instruction with no demonstrations: the combination the search selected on validation uses none.

The saved program was compiled without the locked test: the optimizer saw training and validation data only, and the locked test was scored once afterwards.

> **Book vs. notebook:** The recorded run fixed `seed=42` and left the evaluation thread count at DSPy's default. The printed listing uses DSPy's default seed and `NUM_THREADS = 1`, so a live run evaluates sequentially, takes roughly 20 minutes to compile instead of the recorded 285 seconds, and can select a different program.

The next cell previews the saved program.

In [7]:
print(learned_program_preview(OPTIMIZER))
print()
print("Saved program matches saved prompt:", verify_prompt_artifact(OPTIMIZER))

Predictor: detect.predict
Learned instruction (1807 characters):
You are a forensic editor specializing in technical documentation and AI-rewrite detection. Decide whether the supplied passage is likely AI-generated (`True`) or human-authored (`False`) by focusing on stylistic rewrite artifacts rather than its topic, terminology, correctness, or technical sophistication.

AI rewrites in this benchmark often preserve the original meaning while making the prose more editorially polished and structurally elaborate. Look for clusters of signals such as:
- unnecessary clarifications, parenthetical explanations, or interpretive commentary;
- explicit causal or logical links where concise documentation would leave them implicit;
- generalized, elevated, formulaic, or overly balanced wording;
- slogan-like restatements or repetition of the same point in refined language;
- unusually smooth transitions, symmetrical sentence construction, or elaborate punctuation;
- expansion of simple guidance 

## Apply the pattern

Adapt the listing above to your own DSPy program, metric, and frozen train/validation split. Keep the test set untouched until the optimizer returns, then report locked-test accuracy as `correct / test examples` so every optimizer is easy to compare. Use the baseline notebook (`quickstart-ai-detector.ipynb`) when you also need uplift.

All Chapter 6 results are summarized in `CHAPTER_RESULTS.md`. Scores, prompts, programs, and predictions are under `results/expanded_notebooks/` (GEPA: `results/gepa_light_standard/`). Fine-tuned model weights are generated locally and are not stored in the repository. No credentials are stored.